doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+# Code-generator
+
+This guide focuses on writing new templates for the code-generator. For using the code generator take a look at the man-page
+[`kdb-gen(1)`](/doc/help/kdb-gen.md). For specifically using the code-generator for the high-level API you may want to follow
+[this guide](/doc/tutorials/highlevel.md).
+
+## Basics
+
+The underlying framework of `kdb gen` is quite flexible. It is based on the [mustache templating system](https://mustache.github.io/).
+Concretely we use [this C++ library](https://github.com/kainjow/Mustache) as a basis for our framework.
+
+The file [`src/tools/kdb/gen.cpp`](/src/tools/kdb/gen.cpp) implements the command-line interface (CLI) and is of little interest to this
+guide. Instead we will focus on the framework that is invoked via the CLI. The bulk of the framework is implemented in the classes
+[`GenTemplate` and `GenTemplateList`](/src/tools/kdb/gen/template.hpp) (and of course kainjow's mustache library).
+
+First we need to define a few terms:
+
+- _template name_: This is the main identifier of a template. It is used in the command line to choose the template.
+- _template base name_: Unlike the template name, this identifier is purely internal. It is the base name of all mustache files belonging
+  to the template. In most cases this will be the same as the template name.
+- _input keyset_: This keyset contains the data with which the template will be instantiated by the code-generator.
+- _parent key_: The parent key is given in the command line and defines the input keyset (the keys below the parent key).
+- _output name_: The base name for the output files as given in the command line. Suffixes may be appended, if there are multiple output
+  files.
+- _parts_: Each template may consist of multiple parts. Each part corresponds to exactly one mustache file and therefore exactly one output
+  file.
+- _part suffix_: Parts are identified by there part suffix. This suffix is appended to the template base name to obtain the name of the
+  mustache file corresponding to the part. Similarly it is appended to the output name to obtain the output file for this part. The empty
+  string `""` is a valid part suffix, although it is only recommend for templates with a single output file.
+
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing. The file will simple list all keys in our input
+keyset. An example output would be (the same as `kdb ls`):
```

Comment:
output of the template engine or of an execution of the binary from the compiled code we got from the template engine?

Reply:
There is no compiled binary. We do not create a template for C code, but just a very basic text file template to explain how the code-generator framework works.